src/FFmpegWriter.cpp

```diff
@@ -102,6 +102,34 @@ FFmpegWriter::FFmpegWriter(std::string path) :
 	auto_detect_format();
 }
 
+FFmpegWriter::~FFmpegWriter() {
+	if (is_open) {
+		Close();
```

Comment:
I've only just taken a cursory look at this, but one thing I will say: If _at all_ possible, it would be better to avoid calling `Close()` in the destructor. There's too much of that already in libopenshot, and we need to avoid it because our `Close()` methods are _waaaay_ too heavy. They do a lot of housekeeping "stuff" in preparation for being **reopened** again, which of course makes no sense in a destructor.

See the discussion at #378, and in particular https://github.com/OpenShot/libopenshot/issues/378#issuecomment-559572714, for more on that.

It may mean having to reimplement some parts of `Close()` in the destructor, or separate them out into a new method (`Finalize()`?) that `Close()` also calls, but it would be good to avoid all of the unnecessary "stuff" in `Close()`.

Reply:
Hmm. FFmpegWriter's `Close()` is actually pretty good, as libopenshot `Close()` methods go. It doesn't do too much rearranging of deck chairs on the Titanic, as the saying goes, and most of what it _does_ is properly conditional.

It does _unconditionally_ do this, though:
https://github.com/OpenShot/libopenshot/blob/49972b21e9b5f3cab08d7882b93c1f37dc6577ac/src/FFmpegWriter.cpp#L1020-L1022

...which may be a problem in a destructor, if there is no valid `oc` to free. May be as simple as slapping an `if (oc)` around it.

Reply:
It also calls ZeroMQ as its very last action:
https://github.com/OpenShot/libopenshot/blob/49972b21e9b5f3cab08d7882b93c1f37dc6577ac/src/FFmpegWriter.cpp#L1030

which DEFINITELY should not be done from a destructor. But since that's just a debug `printf()` with no useful information, really it can just be removed.

Reply:
Actually, I take it back. `Close()` itself is fairly lightweight, but the methods it calls are WAY too heavy. `WriteTrailer()`, `close_video()`, `close_audio()`, they all do a **lot** of processing that will bite us in a destructor.

If the video file hasn't been successfully written by the time the destructor is called, I don't think it's realistic for the destructor to try and _complete_ the encoding process before it goes away. That's what an explicit `Close()` is for, and if the caller hasn't called it by the time the FFmpegWriter object is being _destroyed_, IMHO it should just close the output stream and drop everything on the floor, not try and write it all out first.